### PR TITLE
ship a better storage solution

### LINKED ISSUE #1151: Ship a working storageclass
**What would you like to be added**:

We would like KIND to create a working storageclass by default.

**Why is this needed**:

We build a few Kubernetes operators. We would like to update our documentation so that users use KIND instead of Minikube as a test/dev Kubernetes setup in case they don't have one. Here is an example: https://github.com/stashed/docs/pull/54

One of the blockers for us is that KIND does not come with a working storageclass. We had success with Rancher's local provisioner. This is what we do internally. https://github.com/appscodelabs/e2e-test-setup/blob/kind-v0.6.1/run.sh

Obviously we could point our users to our custom docs for how to setup KIND, but it will be nice it KIND could ship a working storageclass solution.

### PR DESCRIPTION
**Pull request by BenTheElder**
- install a customized rancher.io/local-host-path driver config at image build time (with preloaded images)
- when installing the default storage class, attempt to install from the node instead of from the one in the kind binary first

TODO:
- [x] handle kubernetes < 1.12
  - DONE: for less than 1.12 we will ship the old manifest.
- [x] Push an image

I'll handle those later tonight. I just want to surface progress on this.

Fixes #1151 #118

### CODE REVIEW

**Comment by BenTheElder** [comment]
note that this does unfortunately increase the node image size slightly, but having somewhat saner PVs is worth it.

File: pkg/build/node/node.go [hunk 1]
```diff
@@ -447,6 +436,8 @@ func (c *BuildContext) prePullImages(dir, containerID string) error {
 
 	// all builds should isntall the default CNI images currently
 	requiredImages = append(requiredImages, defaultCNIImages...)
+	// all builds should isntall the default storage driver images currently
```
Comment (aojea) [inline pkg/build/node/node.go]: carrying over the typo 🙃

File: pkg/build/node/storage.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+metadata:
+  name: local-path-config
+  namespace: local-path-storage
+data:
+  config.json: |-
+        {
+                "nodePathMap":[
+                {
+                        "node":"DEFAULT_PATH_FOR_NON_LISTED_NODES",
+                        "paths":["/var/local-path-provisioner"]
```
Comment (aojea) [inline pkg/build/node/storage.go]: /var is a volume if IIRC, does this have any implication?

Comment (BenTheElder) [reply pkg/build/node/storage.go]: Yes, that's why we selected it. It's the appropriate place for persistent data.

Comment (BenTheElder) [reply pkg/build/node/storage.go]: There is a set of notes about the delta above.

File: pkg/cluster/internal/create/actions/installstorage/storage.go [hunk 3]
```diff
@@ -53,7 +55,7 @@ func (a *action) Execute(ctx *actions.ActionContext) error {
 	node := controlPlanes[0] // kind expects at least one always
 
 	// add the default storage class
-	if err := addDefaultStorageClass(node); err != nil {
+	if err := addDefaultStorage(ctx.Logger, node); err != nil {
```
Comment (aojea) [inline pkg/cluster/internal/create/actions/installstorage/storage.go]: should we be able to opt-out and not install it as we did with the CNI?
or opt-in and keep the legacy by default?

Comment (BenTheElder) [reply pkg/cluster/internal/create/actions/installstorage/storage.go]: Not necessary initially.
You can easily replace the storage class with your own and delete the deployment, unlike CNI.

Let me mark this PR WIP.

Comment (BenTheElder) [reply pkg/cluster/internal/create/actions/installstorage/storage.go]: There's no point in opt-in. We ship hermetic images, the images etc. will already be there and just running the deployment is low cost.

And the legacy one is know to have major limitations.

File: pkg/build/node/node.go (hunk 1 above)
Comment (BenTheElder) [reply pkg/build/node/node.go]: fixed

**Comment by BenTheElder** [comment]
handled v1.11 node image builds by shipping the old manifest in that case (and not bundling the images)

that's pretty low effort but will let us continue to best effort support v1.11 minimally without much code

**Comment by BenTheElder** [comment]
rebased to upstream and pushed the node image

**Comment by aojea** [comment]
/lgtm 
I couldn't test it though, will be nice to have feedback from users

**Comment by tamalsaha** [comment]
This is great! Looking forward to the next release of Kind.

**Comment by BenTheElder** [comment]
I did a _LOT_ of testing with these. It should work great with the tweaks we've applied for kind :-)

**Comment by monotek** [comment]
Is there a timeline / roadmap / due date for the next kind release?

Very interested in this fix as i hit this with the Elasticsearch (https://github.com/elastic/helm-charts/issues/429) and likely also the Prometheus chart (https://github.com/eclipse/packages/pull/13#issuecomment-573674567).

**Comment by BenTheElder** [comment]
https://github.com/kubernetes-sigs/kind/issues/1250 working on it

nominally we use milestones, but that got a bit out of date over the
holidays.

**Comment by BenTheElder** [comment]
the release is out!

**Comment by monotek** [comment]
Thanks a lot! :-)

**Pull request merged**